staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -0,0 +1,369 @@
+			response.locallyServiceable = true
+			return false
+		}
+		serviceableBy = append(serviceableBy, apiserverKey)
+		return true
+	})
+	response.serviceableBy = serviceableBy
+	return response, nil
+}
+
+func (h *peerProxyHandler) proxyRequestToDestinationAPIServer(req *http.Request, crw *CustomResponseWriter, host string) {
+	user, ok := apirequest.UserFrom(req.Context())
+	if !ok {
+		klog.Errorf("failed to get user info from request")
+		return
+	}
+
+	// write a new location based on the existing request pointed at the target service
+	location := &url.URL{}
+	location.Scheme = "https"
+	location.Host = host
+	location.Path = req.URL.Path
+	location.RawQuery = req.URL.Query().Encode()
+
+	newReq, cancelFn := apiserverproxyutil.NewRequestForProxy(location, req)
+	newReq.Header.Add(PeerProxiedHeader, "true")
+	defer cancelFn()
+
+	// create transport
+	clientConfig := &transport.Config{
+		TLS: transport.TLSConfig{
+			Insecure:   false,
+			CertFile:   h.proxyClientCertFile,
+			KeyFile:    h.proxyClientKeyFile,
+			CAFile:     h.peerCAFile,
+			ServerName: "kubernetes.default.svc",
+		},
+	}
+
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	defer cancelCtx()
+	proxyRoundTripper, transportBuildingError := transport.NewWithContext(ctx.Done(), clientConfig)
+	if transportBuildingError != nil {
+		klog.Error(transportBuildingError.Error())
+		return
+	}
+
+	proxyRoundTripper = transport.NewAuthProxyRoundTripper(user.GetName(), user.GetGroups(), nil, proxyRoundTripper)
+
+	handler := proxy.NewUpgradeAwareHandler(location, proxyRoundTripper, true, false, &responder{w: crw, ctx: req.Context()})
+	handler.ServeHTTP(crw, newReq)
+	// Increment the count of successful proxied requests
+	metrics.IncPeerProxiedRequest(ctx, strconv.Itoa(crw.statusCode))
+}
+
+// responder implements rest.Responder for assisting a connector in writing objects or errors.
+type responder struct {
+	w   http.ResponseWriter
+	ctx context.Context
+}
+
+func (crw *CustomResponseWriter) WriteHeader(code int) {
+	crw.statusCode = code
+	crw.ResponseWriter.WriteHeader(code)
+}
+
+func (r *responder) Error(w http.ResponseWriter, req *http.Request, err error) {
+	klog.Errorf("Error while proxying request to destination apiserver: %v", err)
+	// Increment the count of failed proxied requests
+	http.Error(w, err.Error(), http.StatusServiceUnavailable)
+}
+
+// Adds a storageversion object to SVMap
+func (h *peerProxyHandler) addSV(obj interface{}) {
+	sv := obj.(*v1alpha1.StorageVersion)
+	h.updateSVMap(nil, sv)
+}
+
+// Updates the SVMap to delete old storageversion and add new storageversion
+func (h *peerProxyHandler) updateSV(oldObj interface{}, newObj interface{}) {
+	oldSV := oldObj.(*v1alpha1.StorageVersion)
+	newSV := newObj.(*v1alpha1.StorageVersion)
+	h.updateSVMap(oldSV, newSV)
+}
+
+// Deletes a storageversion object from SVMap
+func (h *peerProxyHandler) deleteSV(obj interface{}) {
+	sv := obj.(*v1alpha1.StorageVersion)
```

Comment:
Can `obj` be a https://pkg.go.dev/k8s.io/client-go/tools/cache#DeletedFinalStateUnknown?

Reply:
added a check to verify the final state of obj